react/src/component/login/login.tsx

```diff
@@ -0,0 +1,217 @@
+    )
+
+
+}
+
+export const CatchLoggedOut = (props) => {
+    let appName = useContext(CAppName);
+    let loginStatus = useContext(CLoginStatus);
+
+    return (
+        <>
+            {
+                (loginStatus.isLoggedIn && !loginStatus.needCompleteRegistration) || loginStatus.method === "guest" ?
+                (
+                    props.children
+                ) : (
+                    <Navigate to={`/react/${appName}/login`}/> // TODO @garsuga: abstract
+                )
+            }
+        </>
+    )
+}
+
+type LoginConfirmParams = {
+    token: string,
+    method: AuthMethod,
+    needCompleteRegistration: "0" | "1"
+}
+
+export const LoginConfirm = (props) => {
+    let { token, method, needCompleteRegistration } = useParams<LoginConfirmParams>();
+
+    switch(method) {
+        case 'email':
+            return <LoginEmailConfirm needCompleteRegistration={needCompleteRegistration} token={token}/>
+        default:
+            throw new Error(`could not handle login method=${method}`)
+    }
+}
+
+
+
+export const LoginNeedCompleteRegistration = (props) => {
+    let appName = useContext(CAppName);
+    let navigate = useNavigate();
+    let onCompleteForm = (data: {[key: string]: any}) => {
+        fetch(`/auth-complete-registration`, {
+            method: "POST",
+            body: JSON.stringify({ ...data, simulationType: appName }),
+            headers: {
+                "Content-Type": "application/json"
+            }
+        }).then(() => navigate(`/react/${appName}`))
+    }
+    return (
+        <LoginExtraInfoForm onComplete={(data) => onCompleteForm(data)}/>
+    )
+}
+
+export const LoginExtraInfoForm = (props: { onComplete: ({displayName}) => void }) => {
+    let { onComplete } = props;
+    let [displayName, updateDisplayName] = useState<string>("");
+    return (
+        <>
+            <p>Please enter your full name to complete your Sirepo registration.</p>
+            <Form.Group as={Row} className="mb-3" controlId="formPlaintextEmail">
+                <Form.Label column sm="3">
+                    Your full name
+                </Form.Label>
+                <Col sm="9">
+                    <Form.Control value={displayName} onChange={(e) => updateDisplayName(e.target.value)}/>
+                </Col>
+            </Form.Group>
+            <Button variant="primary" onClick={() => onComplete({ displayName })}>Continue</Button>
+        </>
+    )
+}
+
+export const LoginRoot = (props) => {
+    let loginStatus = useContext(CLoginStatus);
+    let appName = useContext(CAppName);
+
+    let getLoginComponent = (method: string): JSX.Element => {
+        switch(method) {
+            case "email":
+                return <LoginWithEmail/>
+            case "guest":
+                return <Navigate to={new AppWrapper(appName).getAppRootLink()}/>
+            default:
+                throw new Error(`could not handle login method=${method}`)
```

Comment:
"could not handle" > "unknown". Try to keep error messages tokenizable and short. You could even code them as unknown-login-method, which can then be parsed on the server side.